src/test/clone/test_fork.rs

```diff
@@ -383,6 +384,537 @@ fn test_exit_signal_with_fatal_signal(exit_signal: nix::sys::signal::Signal) ->
+) -> anyhow::Result<()> {
+    run_test_in_subprocess(|| {
+        // `waitid(2)`: A "clone" child is  one  which delivers  no signal, or a
+        // signal other than SIGCHLD to its parent upon termination.
+
+        let clone_child_no_signal = match unsafe {
+            linux_api::sched::clone(
+                CloneFlags::empty(),
+                None,
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+            )
+        }
+        .unwrap()
+        {
+            CloneResult::CallerIsChild => {
+                unsafe { libc::exit(0) };
+            }
+            CloneResult::CallerIsParent(child_pid) => child_pid,
+        };
+        let clone_child_alt_signal = match unsafe {
+            linux_api::sched::clone(
+                CloneFlags::empty(),
+                Some(Signal::SIGURG),
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+            )
+        }
+        .unwrap()
+        {
+            CloneResult::CallerIsChild => {
+                unsafe { libc::exit(0) };
+            }
+            CloneResult::CallerIsParent(child_pid) => child_pid,
+        };
+        let reg_child = match unsafe { linux_api::sched::fork() }.unwrap() {
+            CloneResult::CallerIsChild => {
+                unsafe { libc::exit(0) };
+            }
+            CloneResult::CallerIsParent(child_pid) => child_pid,
+        };
+
+        let options = if use_wclone { libc::__WCLONE } else { 0 };
+        let results = HashSet::from([waitfn(options), waitfn(options), waitfn(options)]);
+        let expected = if use_wclone {
+            HashSet::from([
+                Some(clone_child_no_signal),
+                Some(clone_child_alt_signal),
+                None,
+            ])
+        } else {
+            HashSet::from([Some(reg_child), None, None])
+        };
+        assert_eq!(results, expected);
+    })
+}
+
+fn test_waitfn_honors_wall(
+    use_wall: bool,
+    waitfn: impl Fn(/*options*/ i32) -> Option<Pid>,
+) -> anyhow::Result<()> {
+    run_test_in_subprocess(|| {
+        // `waitid(2)`: A "clone" child is  one  which delivers  no signal, or a
+        // signal other than SIGCHLD to its parent upon termination.
+
+        let clone_child_no_signal = match unsafe {
+            linux_api::sched::clone(
+                CloneFlags::empty(),
+                None,
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+                core::ptr::null_mut(),
+            )
+        }
+        .unwrap()
+        {
+            CloneResult::CallerIsChild => {
+                unsafe { libc::exit(0) };
+            }
+            CloneResult::CallerIsParent(child_pid) => child_pid,
+        };
```

Comment:
nit: maybe split up some of these to make them easier to read? for example maybe `let clone_rv = linux_api::sched::clone(...); let clone_child_no_signal = match clone_rv { ... };`

Reply:
done